File: yologram-api-v2/app/infra/cache/tech_news_first_page_cache.py
from collections.abc import Callable

from app.core.response import ApiEnvelopCursorPage
from app.domain.news.tech.schema import TechNewsResponse
from app.infra.cache.cache import Cache
from app.infra.cache.cache_service import CacheService
from app.infra.cache.redis_cache_service import RedisCacheService


class TechNewsFirstPageCache:
    """
    테크 뉴스 첫 페이지 cache-aside 공용 컴포넌트 (api-v1 TechNewsFirstPageCache 미러).
    조회는 페이지 키 GET 1회 — 버전 키 간접층 없음. 무효화는 worker가 새 뉴스 요약 시
    첫 페이지 키를 전수 열거해 UNLINK(worker TechNewsFirstPageCacheInvalidator와 키 계약).
    TTL 3분은 삭제 누락·레이스(삭제 직후 옛 목록 SET 부활) 대비 보험이자 낡음의 상한.
    저장 값은 응답 envelope JSON(camelCase, datetime ISO) — api-v1 Jackson이 직렬화한
    ApiEnvelopCursorPage<TechNewsResponse>와 동일 구조라 v1↔v2 캐시 상호 호환.
    DB 조회는 loader 콜러블로 주입받아 도메인 경계(리포지토리 소유)는 호출부에 남긴다.
    """

    def __init__(self, cache_service: CacheService | None = None):
        self.cache_service: CacheService = cache_service or RedisCacheService()

    def get(
        self,
        category_id: int | None,
        size: int,
        loader: Callable[[], ApiEnvelopCursorPage[TechNewsResponse]],
    ) -> ApiEnvelopCursorPage[TechNewsResponse]:
        """첫 페이지 cache-aside: 페이지 키 조회 → 히트 복원 / 미스면 loader → 저장."""
        page_cache = Cache.tech_news_first_page(category_id, size)
        cached = self.cache_service.get_or_null(page_cache)
        if cached is not None:
            return self._from_cached(cached)

        loaded = loader()
        self.cache_service.set(page_cache, self._to_cacheable(loaded))
        return loaded

    @staticmethod
    def _to_cacheable(page: ApiEnvelopCursorPage[TechNewsResponse]) -> dict:
        # camelCase(by_alias) + datetime ISO 초 단위(mode="json") — v1 LocalDateTime 직렬화와 일치.
        # null nextCursor는 envelope의 model_serializer가 생략 (v1 @JsonInclude NON_NULL 정합)
        return page.model_dump(by_alias=True, mode="json")

    @staticmethod
    def _from_cached(cached: dict) -> ApiEnvelopCursorPage[TechNewsResponse]:
        data = [TechNewsResponse.model_validate(item) for item in cached.get("data", [])]
        # nextCursor 미존재(null 생략 규약) → None
        return ApiEnvelopCursorPage(data=data, next_cursor=cached.get("nextCursor"))

File: yologram-api-v1/src/main/kotlin/link/yologram/api/v1/infra/cache/TechNewsFirstPageCache.kt
package link.yologram.api.v1.infra.cache

import link.yologram.api.v1.domain.news.tech.model.TechNewsResponse
import link.yologram.api.v1.global.model.ApiEnvelopCursorPage
import org.springframework.stereotype.Component

/**
 * 테크 뉴스 첫 페이지 cache-aside 공용 컴포넌트.
 *
 * - 조회는 페이지 키 GET 1회 — 버전 키 간접층 없음.
 * - 무효화: worker가 새 뉴스 요약(SUMMARIZED) 시 첫 페이지 키를 전수 열거해 UNLINK —
 *   다음 요청부터 미스가 나 최신 목록이 다시 캐시된다 (worker TechNewsFirstPageCacheInvalidator와 키 계약).
 * - TTL 3분은 보험: 삭제 실패(Redis 순단·worker 다운)나 레이스(삭제 직후, 커밋 전에 읽은
 *   옛 목록이 SET으로 부활)로 낡은 캐시가 남아도 3분이 생존 상한.
 * - 첫 페이지(cursor == null)만 대상 — 커서 페이지는 키가 커서 값으로 분산돼 히트율이 없어 제외.
 * - Redis 장애 시 CacheService가 실패를 삼켜 미스와 동일 경로 → loader(DB) 폴백, 기능 무손상.
 *
 * UserNicknameCache와 동일하게 DB 조회는 loader 람다로 주입받아
 * 도메인 경계(리포지토리·클라이언트 소유)는 도메인 서비스에 남긴다.
 */
@Component
class TechNewsFirstPageCache(
    private val cacheService: CacheService,
) {

    /** cache-aside: 페이지 키 히트 시 반환, 미스 시 loader → 결과 캐시 후 반환. */
    fun get(
        categoryId: Long?,
        size: Int,
        loader: () -> ApiEnvelopCursorPage<TechNewsResponse>,
    ): ApiEnvelopCursorPage<TechNewsResponse> {
        val cache = Cache.techNewsFirstPage(categoryId, size)
        cacheService.getOrNull(cache)?.let { return it }

        val loaded = loader()
        cacheService.set(cache, loaded)
        return loaded
    }
}
